### PR TITLE
first pass to image chat attachments

### PR DESCRIPTION
**Pull request by justschen**
allows users to attach from clipboard (via paste or quick pick), or drag and drop an image into chat to be used as context. copilot does not yet support, but purely from a UI perspective.

on web, there will be this pop up:
Screenshot 2024-09-17 at 2 31 28 PM

### CODE REVIEW

File: src/vs/platform/clipboard/browser/clipboardService.ts
Comment (roblourens) [inline src/vs/platform/clipboard/browser/clipboardService.ts:108]: Looks like this isn't called

File: src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts:323]: Should be able to remove this if you aren't registering a variable from the extension anymore

File: src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts:232]: In the scenario where the image is a file on disk, I think I would expect it to work just like 'file' where the value is a URI, and the extension reads the file contents. In a remote window, for example, this would spend time copying the file from the remote EH to the window and back.

File: src/vs/workbench/contrib/chat/browser/chatImagePaste.ts
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/chatImagePaste.ts:54]: Does OpenAI support more image types?

File: src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts
Comment (justschen) [reply src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts:232]: does that mean we want the extension to do the conversion to `base64`? I did this so that the `IChatRequestVariableEntry` would already have image data, labels, id, etc.

File: src/vs/workbench/contrib/chat/browser/chatImagePaste.ts
Comment (justschen) [reply src/vs/workbench/contrib/chat/browser/chatImagePaste.ts:54]: added the remaining image types (gif and tiff)

File: src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts
Comment (justschen) [reply src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts:323]: I'm still creating an `imageResolver` on the extension side since i saw that most of the attachments were being attached to the quick pick in the extension. could this be done elsewhere, or would this be more in line with what we're already doing?

Comment (justschen) [reply src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts:323]: jk this is a dumb question - directly adding it to quick pick here is fine as well i think

File: src/vs/workbench/contrib/chat/browser/chatDragAndDrop.ts
Comment (benibenj) [inline src/vs/workbench/contrib/chat/browser/chatDragAndDrop.ts:137]: If we keep images part of the file drop type, I would move this if/else logic into `getEditorAttachContext`. I didn't pick a great name for the function, maybe we need to change it to something like `getContextFromInput`

File: src/vs/workbench/contrib/chat/browser/chatDragAndDrop.ts
Comment (benibenj) [inline src/vs/workbench/contrib/chat/browser/chatDragAndDrop.ts:201]: I think `id` needs to be a unique identifier for the specific image. This then allows us to make sure we only attach the same image once. Probably `id: resource.toString()` would work.

File: src/vs/workbench/contrib/chat/browser/chatDragAndDrop.ts
Comment (benibenj) [inline src/vs/workbench/contrib/chat/browser/chatDragAndDrop.ts:135]: A bit suprised to see a `readFile` call in here. Would it not be possible to pass the resource as attachement like we do with files and have it be read later on?

Comment (justschen) [reply src/vs/workbench/contrib/chat/browser/chatDragAndDrop.ts:135]: https://github.com/microsoft/vscode/pull/228706#discussion_r1762260699

i guess kind of similar to this question here. since we are creating the attachment, would we want to create the attachment with the correct type and id and such (for image, it will be the buffer/uint8array) or have the extension convert it? i'm not sure what is best

File: src/vs/workbench/contrib/chat/browser/chatDragAndDrop.ts
Comment (justschen) [reply src/vs/workbench/contrib/chat/browser/chatDragAndDrop.ts:131]: I think in the `DataTransferItem`, we can actually see the data type (for image case, it is `image/png` or such). I'll see what i can do

Screenshot 2024-09-17 at 11 03 29 AM

File: src/vs/workbench/contrib/chat/browser/chatDragAndDrop.ts
Comment (justschen) [reply src/vs/workbench/contrib/chat/browser/chatDragAndDrop.ts:209]: `isFile` is checked later on for the ui of the attachment in `chatInputPart` I think it should be `false`, and I added a `isImage` to better source the attachment

File: src/vs/workbench/contrib/chat/browser/chatInputPart.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/chat/browser/chatInputPart.ts:631]: Prefer `addDisposableListener` so you can register and manage its lifecycle.

File: src/vs/workbench/contrib/chat/browser/chatInputPart.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/chat/browser/chatInputPart.ts:634]: You can do this in a more strongly typed way via `StandardKeyboardEvent` https://github.com/microsoft/vscode/blob/6c478a6b044c36a48a3559213f9f8d0460831612/src/vs/base/browser/dom.ts#L2027-L2046

File: src/vs/workbench/contrib/chat/browser/chatInputPart.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/chat/browser/chatInputPart.ts:603]: No need to localize this as it's just the given name

File: src/vs/workbench/contrib/chat/browser/chatInputPart.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/chat/browser/chatInputPart.ts:605]: A span shouldn't have a tabIndex

File: src/vs/workbench/contrib/chat/browser/chatInputPart.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/chat/browser/chatInputPart.ts:596]: This alt text is actually worse than nothing, if we can't describe the image it's best to leave this as `alt=""` so the screen reader treats it as decorative

File: src/vs/workbench/contrib/chat/browser/chatInputPart.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/chat/browser/chatInputPart.ts:587]: Same comment as below wrt alt text

File: src/vs/workbench/contrib/chat/browser/chatInputPart.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/chat/browser/chatInputPart.ts:617]: We should remove the one we don't want here

File: src/vs/workbench/contrib/chat/browser/chatImagePaste.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/chat/browser/chatImagePaste.ts:60]: These image signatures/magic numbers? An explanation/short description of what they are would be good for those uninitiated in the topic. Also png's full signature is `89 50 4E 47 0D 0A 1A 0A`

File: src/vs/workbench/contrib/chat/browser/chatImagePaste.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/chat/browser/chatImagePaste.ts:24]: Capital D? Also to avoid naming conflicts with events I normally call these something like `_handlePaste`

File: src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts:62]: Why is `name` hardcoded? I see a check below but shouldn't that be checking `kind` instead?

File: src/vs/platform/clipboard/common/clipboardService.ts
Comment (Tyriar) [inline src/vs/platform/clipboard/common/clipboardService.ts:57]: Some jsdoc would be helpful here. Would be handy to mention if this causes a permissions dialog on web.

File: src/vs/platform/clipboard/browser/clipboardService.ts
Comment (Tyriar) [inline src/vs/platform/clipboard/browser/clipboardService.ts:53]: We support a few types below but here is only png?

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts:230]: Little confused by what images we support where, there is png above, png+jpg here, png+jpg+bmp+gif+tiff below?

File: src/vs/workbench/contrib/chat/browser/chatDragAndDrop.ts
Comment (benibenj) [reply src/vs/workbench/contrib/chat/browser/chatDragAndDrop.ts:139]: I'm not sure to be honest. I think ideally it would be the same as text files. We don't read those and save the contents in `value` so it should also not be done for images in theory. However, I assume this would not work for dnd from the web? If that is the case then supporting two different ways of image contexts would also not be great...

File: src/vs/workbench/contrib/chat/browser/chatInputPart.ts
Comment (benibenj) [inline src/vs/workbench/contrib/chat/browser/chatInputPart.ts:625]: What about having the hover show as soon as the user focuses on it? This is what we normally do. I doubt users will figure out that they need to focus the item and press `Enter` or `Space`.

Comment (justschen) [reply src/vs/workbench/contrib/chat/browser/chatInputPart.ts:625]: the `setUpManagedHover` does this for us (it will show hover after a slight delay). I was wondering if we should show immediately vs. having delay, but most hovers it seems are shown after a delay.

Comment (benibenj) [reply src/vs/workbench/contrib/chat/browser/chatInputPart.ts:625]: We don't recommend removing the delay at the start as it can be pretty annoying in some cases. However, what I recommend is to create an instant hover delegate (`createInstantHoverDelegate()` to create, but needs to be disposed) outside this loop and use the same instance for all managed hovers. This would allow you to tab through the attachments and have the hovers show instantly after having to wait the delay the first time.

File: src/vs/workbench/contrib/chat/browser/chatDragAndDrop.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/chat/browser/chatDragAndDrop.ts:138]: gif typo

File: src/vs/workbench/contrib/chat/browser/chatInputPart.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/chat/browser/chatInputPart.ts:593]: Not needed? I think what you're after is to have the container be focusable, is that `widget`?

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/platform/clipboard/common/clipboardService.ts
Comment (Tyriar) [inline src/vs/platform/clipboard/common/clipboardService.ts:60]:
```suggestion
	/**
	 * Reads resources from the system clipboard as an image. If the clipboard does not contain an
	 * image, an empty buffer is returned.
	 */
```

**Review by Tyriar — CHANGES_REQUESTED** [review]

**Review by Tyriar — DISMISSED** [review]

File: src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts:276]: Are these shown to the user and need to be localized?

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts
Comment (Tyriar) [inline src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts:276]: `fullName` too? I'm not sure the difference

Comment (justschen) [reply src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts:276]: `fullName` doesn't get used anywhere else, but we do show `attachment.name` as a label

File: src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts:334]: localize

File: src/vs/workbench/contrib/chat/browser/chatInputPart.ts
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/chatInputPart.ts:654]: You can pass a callback so the content doesn't have to be computed until the hover happens. Can you do it that way and avoid reading the file immediately? As is, it will read all the files from disk for every attachment, every time an attachment is added or removed, and that's unnecessary work. I also don't like this loop being async. There's nothing preventing `initAttachedContext` from running twice and having two async loops interfering with each other.

File: src/vs/workbench/contrib/chat/browser/chatInputPart.ts
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/chatInputPart.ts:618]: Using the `dom.$` or `dom.h` helpers is much cleaner. You don't necessarily have to change this but you should look into it for next time, it will save a lot of lines of boilerplate.

File: src/vs/workbench/contrib/chat/browser/chatImagePaste.ts
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/chatImagePaste.ts:51]: Are you sure this is unique enough?

**Review by roblourens — COMMENTED** [review]
Do you plan to merge this? Does something prevent it from showing up now while we don't have image support in models?

File: src/vs/workbench/contrib/chat/browser/chatInputPart.ts
Comment (roblourens) [reply src/vs/workbench/contrib/chat/browser/chatInputPart.ts:654]: This shouldn't be `this._register`, it should go on this.attachedContextDisposables

File: src/vs/workbench/contrib/chat/browser/chatImagePaste.ts
Comment (justschen) [reply src/vs/workbench/contrib/chat/browser/chatImagePaste.ts:51]: I tested with 12 bytes (which is technically the max bytes of identifiers for PNGs, every other file type is less than that), and it had some trouble in similar sized, similar screenshots up until around (0,20). i think 50 is okay, but open to make the identifier longer just in case

File: src/vs/platform/clipboard/browser/clipboardService.ts
Comment (benibenj) [inline src/vs/platform/clipboard/browser/clipboardService.ts:54]: Not a blocker for merging but this does not work on Firefox as they do not set these drag types. You can read about it here: https://developer.mozilla.org/en-US/docs/Web/API/HTML_Drag_and_Drop_API/Recommended_drag_types#dragging_images

File: src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts
Comment (Tyriar) [reply src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts:290]: It looks like it's intent is to be shown to the user:

https://github.com/microsoft/vscode/blob/3ae6c5ef651585b12870828230f214eed2e6ef07/src/vs/workbench/contrib/chat/browser/chatInputPart.ts#L641

I'm guessing this should either match `name` exactly (localized) or be undefined.

Comment (justschen) [reply src/vs/workbench/contrib/chat/browser/actions/chatContextActions.ts]: ope yes forgot about that one 😭

**Review by Tyriar — DISMISSED** [review]
Got some conflicts but lgtm

**Review by Tyriar — DISMISSED** [review]

File: src/vs/workbench/contrib/chat/browser/chat.contribution.ts
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/chat.contribution.ts:133]: I'm not sure the setting should be a registered user-visible setting, because it won't work with current models if they enable it.

File: src/vs/workbench/contrib/chat/browser/chatDragAndDrop.ts
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/chatDragAndDrop.ts:137]: It doesn't look like this needs to be async

File: src/vs/workbench/contrib/chat/browser/chatContentParts/chatAttachmentsContentPart.ts
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/chatContentParts/chatAttachmentsContentPart.ts:97]: You shouldn't have to initialize an empty buffer for this variable, it isn't used

File: src/vs/workbench/contrib/chat/browser/chatInputPart.ts
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/chatInputPart.ts:669]: I guess you didn't start this trend, but we shouldn't be copying/pasting code between the two places that attachments are rendered

File: src/vs/workbench/contrib/chat/browser/chatContentParts/chatAttachmentsContentPart.ts
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/chatContentParts/chatAttachmentsContentPart.ts:46]: I don't know why there's a separate disposable store on this content part, that seems redundant

File: src/vs/workbench/contrib/chat/browser/chatInputPart.ts
Comment (roblourens) [reply src/vs/workbench/contrib/chat/browser/chatInputPart.ts:669]: Can you try to reduce the duplication though? I think the image attachments are a bit more complex and it's just a lot of code to copy.

Comment (roblourens) [reply src/vs/workbench/contrib/chat/browser/chatInputPart.ts:669]: Feel free to do that separately after merging this though.

File: src/vs/workbench/test/electron-sandbox/workbenchTestServices.ts
Comment (roblourens) [inline src/vs/workbench/test/electron-sandbox/workbenchTestServices.ts:159]: This is equivalent to having an async method that returns `Uint8Array.from([])`

**Review by roblourens — DISMISSED** [review]

**Review by roblourens — APPROVED** [review]

**Review by riikolilover — APPROVED** [review]

**Pull request merged**
